cpp/src/arrow/util/hashing.h

```diff
@@ -0,0 +1,805 @@
+      values_.append(static_cast<const char*>(data), length);
+      // Insert hash entry
+      hash_table_.Insert(const_cast<HashTableEntry*>(p.first), h, {memo_index});
+
+      on_not_found(memo_index);
+    }
+    return memo_index;
+  }
+
+  template <typename Func1, typename Func2>
+  int32_t GetOrInsert(const util::string_view& value, Func1&& on_found,
+                      Func2&& on_not_found) {
+    return GetOrInsert(value.data(), static_cast<int32_t>(value.length()),
+                       std::forward<Func1>(on_found), std::forward<Func2>(on_not_found));
+  }
+
+  int32_t GetOrInsert(const void* data, int32_t length) {
+    return GetOrInsert(data, length, [](int32_t i) {}, [](int32_t i) {});
+  }
+
+  int32_t GetOrInsert(const util::string_view& value) {
+    return GetOrInsert(value.data(), static_cast<int32_t>(value.length()));
+  }
+
+  int32_t GetOrInsert(const std::string& value) {
+    return GetOrInsert(value.data(), static_cast<int32_t>(value.length()));
+  }
+
+  // The number of entries in the memo table
+  // (which is also 1 + the largest memo index)
+  int32_t size() const { return static_cast<int32_t>(hash_table_.size()); }
+
+  int32_t values_size() const { return static_cast<int32_t>(values_.size()); }
+
+  const uint8_t* values_data() const {
+    return reinterpret_cast<const uint8_t*>(values_.data());
+  }
+
+  // Copy (n + 1) offsets starting from index `start` into `out_data`
+  template <class Offset>
+  void CopyOffsets(int32_t start, Offset* out_data) const {
+    auto delta = offsets_[start];
+    for (uint32_t i = start; i < offsets_.size(); ++i) {
+      auto adjusted_offset = offsets_[i] - delta;
+      auto cast_offset = static_cast<Offset>(adjusted_offset);
+      assert(static_cast<int32_t>(cast_offset) == adjusted_offset);  // avoid truncation
+      *out_data++ = cast_offset;
+    }
+  }
+
+  template <class Offset>
+  void CopyOffsets(Offset* out_data) const {
+    CopyOffsets(0, out_data);
+  }
+
+  // Copy values starting from index `start` into `out_data`
+  void CopyValues(int32_t start, uint8_t* out_data) const {
+    CopyValues(start, -1, out_data);
+  }
+
+  // Same as above, but check output size in debug mode
+  void CopyValues(int32_t start, int64_t out_size, uint8_t* out_data) const {
+    int32_t offset = offsets_[start];
+    auto length = values_.size() - static_cast<size_t>(offset);
+    if (out_size != -1) {
+      assert(static_cast<int64_t>(length) == out_size);
+    }
+    memcpy(out_data, values_.data() + offset, length);
+  }
+
+  void CopyValues(uint8_t* out_data) const { CopyValues(0, -1, out_data); }
+
+  void CopyValues(int64_t out_size, uint8_t* out_data) const {
+    CopyValues(0, out_size, out_data);
+  }
+
+ protected:
+  struct Payload {
+    int32_t memo_index;
+  };
+  using HashTableType = HashTable<Payload>;
+  using HashTableEntry = typename HashTable<Payload>::Entry;
+  HashTableType hash_table_;
+  std::vector<int32_t> offsets_;
+  std::string values_;
```

Comment:
Hmm: `values_` could grow large; additionally, in a number of places in the codebase, we will want the result of building the hash table to be "released" into an `arrow::BinaryArray` with zero copy. Can this be changed to use `arrow::BufferBuilder` https://github.com/apache/arrow/blob/master/cpp/src/arrow/buffer.h#L372? You'd perhaps want to track the reserved size so that you can call `UnsafeAppend` and skip Status checks (though it may not matter, benchmarks will settle the case)

Reply:
I'll try to.

Reply:
Hmm... the problem is that the hash kernel, as I understand it, is supposed to be reusable. But if we detach the current buffer, the hash table can't be fed anymore (we can't grow it anymore).

Reply:
Let's leave this for follow up work for now. I'll take a closer look

Reply:
(note that reuse of a hash table is already happening, actually, when building delta dictionaries)